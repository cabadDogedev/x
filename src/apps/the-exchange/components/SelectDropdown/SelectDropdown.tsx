// reducer
import { setReceiveChain, setSwapChain } from '../../reducer/theExchangeSlice';

// hooks
import { useAppDispatch, useAppSelector } from '../../hooks/useReducerHooks';

// utils
import { convertChainIdtoName } from '../../utils/converters';

// components
import Body from '../Typography/Body';

// images
import ArrowDown from '../../images/arrow-down.png';
import { ChainType } from '../../utils/types';

type DropdownProps = {
  options: number[];
  onClick: () => void;
  onSelect: () => void;
  isOpen: boolean;
  className?: string;
};

const SelectDropdown = ({ options, onClick, onSelect, isOpen, className }: DropdownProps) => {
  const dispatch = useAppDispatch();
  const swapChain = useAppSelector((state) => state.swap.swapChain as ChainType);
  const receiveChain = useAppSelector((state) => state.swap.receiveChain as ChainType);
  const isSwapOpen = useAppSelector((state) => state.swap.isSwapOpen as boolean);
  const isReceiveOpen = useAppSelector((state) => state.swap.isReceiveOpen as boolean);

  // this will filter the tokens by chain id
  const handleSelectChainId = (option: number) => {
    if (isSwapOpen) {
      dispatch(setSwapChain({chainId: Number(option), chainName: convertChainIdtoName(option)}));
    }
    if (isReceiveOpen) {
      dispatch(setReceiveChain({chainId: Number(option), chainName: convertChainIdtoName(option)}));
    }
    onSelect();
  };

  return (
    <div className={`${className} h-[34px] z-20`}>
      <button
        type="button"
        className="flex justify-between w-full h-full p-2 bg-white rounded-[3px] focus:outline-none"
        onClick={onClick}
      >
        {isOpen && <Body className="text-[#717171]">Select a chain</Body>}
        <Body className="text-black capitalize">{isSwapOpen ? swapChain?.chainName : receiveChain?.chainName}</Body>
        <img src={ArrowDown} />
      </button>
      {isOpen && (
        <div className="relative bg-white border border-t-black rounded-[3px] w-full max-h-[344px] overflow-y-auto capitalize">
          <ul>
            {options.map((option) => (
              <li
                key={option}
                className="px-4 py-2 cursor-pointer hover:bg-gray-100 hover:text-blue-600 border-b last:border-none"
                onClick={() => handleSelectChainId(option)}
              >
                <Body className="border-b-[#8C8C8C]/[.1]">{convertChainIdtoName(option)}</Body>
              </li>
            ))}
          </ul>
        </div>
      )}
    </div>
  );
};

export default SelectDropdown;
