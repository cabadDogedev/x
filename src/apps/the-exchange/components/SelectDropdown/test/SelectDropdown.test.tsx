import renderer, { act } from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';

// provider
import { Provider } from 'react-redux';
import { store } from '../../../../../store';

// reducer
import { 
  setAmountReceive,
  setAmountSwap,
  setBestOffer,
  setIsReceiveOpen,
  setIsSwapOpen,
  setReceiveChain,
  setReceiveToken,
  setReceiveTokenData,
  setSearchTokenResult,
  setSwapChain,
  setSwapToken,
  setSwapTokenData } from '../../../reducer/theExchangeSlice';

// components
import SelectDropdown from '../SelectDropdown';

const mockTokenAssets = [
  { address: '0x01', name: 'Ether', symbol: 'ETH', chainId: 1, decimals: 18, icon: 'iconEth.png' },
  { address: '0x02', name: 'Matic', symbol: 'MATIC', chainId: 137, decimals: 18, icon: 'iconMatic.png' },
];

describe('<SelectDropdown />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    act(() => {
      store.dispatch(setSwapTokenData(mockTokenAssets));
      store.dispatch(setReceiveTokenData(mockTokenAssets));
      store.dispatch(setIsSwapOpen(false));
      store.dispatch(setIsReceiveOpen(false));
      store.dispatch(setSwapChain({ chainId: 1, chainName: 'Ethereum' }));
      store.dispatch(setReceiveChain({ chainId: 137, chainName: 'Polygon' }));
      store.dispatch(setSwapToken(mockTokenAssets[0]));
      store.dispatch(setReceiveToken(mockTokenAssets[1]));
      store.dispatch(setAmountSwap({ tokenAmount: 0.1, usdAmount: 3000 }));
      store.dispatch(setAmountReceive({ tokenAmount: 10, usdAmount: 3000 }));
      store.dispatch(setBestOffer(undefined));
      store.dispatch(setSearchTokenResult([]));
    });
  });

  const options = [1, 137];
  const onClickMock = jest.fn();
  const onSelectMock = jest.fn();

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <SelectDropdown options={options} onClick={onClickMock} onSelect={onSelectMock} isOpen={true} />
        </Provider>
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
  });

  it('renders closed dropdown with correct initial state', () => {
    render(
        <Provider store={store}>
          <SelectDropdown options={options} onClick={onClickMock} onSelect={onSelectMock} isOpen={false} />
        </Provider>
      );

      act(() => {
        store.dispatch(setIsSwapOpen(true));
      });

      expect(screen.getByText('Ethereum')).toBeInTheDocument();
      expect(screen.queryByText('Select a chain')).not.toBeInTheDocument();
  });

  it('renders open dropdown with options when clicked', () => {
    render(
        <Provider store={store}>
          <SelectDropdown options={options} onClick={onClickMock} onSelect={onSelectMock} isOpen={true} />
        </Provider>
      );

      act(() => {
        store.dispatch(setIsSwapOpen(true));
      });

      expect(screen.getByText('Select a chain')).toBeInTheDocument();
      options.forEach((option) => {
        expect(screen.getByText(option)).toBeInTheDocument();
      });
  });

  it('calls onClick when button is clicked', () => {
    render(
        <Provider store={store}>
          <SelectDropdown options={options} onClick={onClickMock} onSelect={onSelectMock} isOpen={false} />
        </Provider>
      );

      act(() => {
        store.dispatch(setIsSwapOpen(true));
      });

      const button = screen.getByRole('button');
      fireEvent.click(button);
      expect(onClickMock).toHaveBeenCalled();
  });

  it('calls onSelect and updates context when an option is selected', () => {
    render(
        <Provider store={store}>
          <SelectDropdown options={options} onClick={onClickMock} onSelect={onSelectMock} isOpen={true} />
        </Provider>
      );

      act(() => {
        store.dispatch(setIsSwapOpen(true));
      });

      const optionElement = screen.getByText('1');
      fireEvent.click(optionElement);

      expect(onSelectMock).toHaveBeenCalled();
      expect(store.getState().swap.swapChain).toEqual({ chainId: 1, chainName: '1' });
  });

  it('handles receive chain selection correctly when isReceiveOpen is true', () => {
    render(
        <Provider store={store}>
          <SelectDropdown options={options} onClick={onClickMock} onSelect={onSelectMock} isOpen={true} />
        </Provider>
      );

      act(() => {
        store.dispatch(setIsReceiveOpen(true));
      });

      const optionElement = screen.getByText('137');
      fireEvent.click(optionElement);

      expect(onSelectMock).toHaveBeenCalled();
      expect(store.getState().swap.receiveChain).toEqual({ chainId: 137, chainName: '137' });
  });

});
