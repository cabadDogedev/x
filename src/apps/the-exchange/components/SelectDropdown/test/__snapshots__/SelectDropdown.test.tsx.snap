// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectDropdown /> renders correctly and matches snapshot 1`] = `
<div
  className="undefined h-[34px] z-20"
>
  <button
    className="flex justify-between w-full h-full p-2 bg-white rounded-[3px] focus:outline-none"
    onClick={[MockFunction]}
    type="button"
  >
    <p
      className="text-base font-medium text-black_grey text-[#717171]"
    >
      Select a chain
    </p>
    <p
      className="text-base font-medium text-black_grey text-black capitalize"
    >
      Polygon
    </p>
    <img
      src="arrow-down.png"
    />
  </button>
  <div
    className="relative bg-white border border-t-black rounded-[3px] w-full max-h-[344px] overflow-y-auto capitalize"
  >
    <ul>
      <li
        className="px-4 py-2 cursor-pointer hover:bg-gray-100 hover:text-blue-600 border-b last:border-none"
        onClick={[Function]}
      >
        <p
          className="text-base font-medium text-black_grey border-b-[#8C8C8C]/[.1]"
        >
          1
        </p>
      </li>
      <li
        className="px-4 py-2 cursor-pointer hover:bg-gray-100 hover:text-blue-600 border-b last:border-none"
        onClick={[Function]}
      >
        <p
          className="text-base font-medium text-black_grey border-b-[#8C8C8C]/[.1]"
        >
          137
        </p>
      </li>
    </ul>
  </div>
</div>
`;
